query-engine/query-engine-c-abi/src/engine.rs

```diff
@@ -0,0 +1,674 @@
+        let json_datasource_overrides =
+            serde_json::from_str(str_datasource_overrides.as_str()).expect("Datasource overrides cannot be parsed");
+        let overrides: Vec<(_, _)> = stringify_env_values(json_datasource_overrides)
+            .unwrap()
+            .into_iter()
+            .collect();
+
+        let datamodel = get_cstr_safe(constructor_options.datamodel).expect("Datamodel must be present");
+        let mut schema = psl::validate(datamodel.into());
+        let config = &mut schema.configuration;
+
+        // let preview_features = config.preview_features();
+
+        schema
+            .diagnostics
+            .to_result()
+            .map_err(|err| ApiError::conversion(err, schema.db.source()))?;
+
+        let base_path = get_cstr_safe(constructor_options.base_path);
+        match &base_path {
+            Some(path) => env::set_current_dir(Path::new(&path)).expect("Could not change directory"),
+            _ => println!("No base path passed!"),
+        }
+
+        config
+            .resolve_datasource_urls_query_engine(
+                &overrides,
+                |key| env.get(key).map(ToString::to_string),
+                // constructor_options.ignore_env_var_errors,
+                true,
+            )
+            .map_err(|err| ApiError::conversion(err, schema.db.source()))?;
+
+        config
+            .validate_that_one_datasource_is_provided()
+            .map_err(|errors| ApiError::conversion(errors, schema.db.source()))?;
+
+        // let enable_tracing = config.preview_features().contains(PreviewFeature::Tracing);
+        // let engine_protocol = constructor_options.engine_protocol.unwrap_or(EngineProtocol::Json);
+        let engine_protocol = EngineProtocol::Json;
+
+        let config_dir_string = get_cstr_safe(constructor_options.native.config_dir).expect("Config dir is expected");
+        let config_dir = PathBuf::from(config_dir_string);
+
+        let builder = EngineBuilder {
+            schema: Arc::new(schema),
+            engine_protocol,
+            native: EngineBuilderNative { config_dir, env },
+        };
+
+        let log_level_string = get_cstr_safe(constructor_options.log_level).unwrap();
+        let log_level = log_level_string.parse::<LevelFilter>().unwrap();
+        let logger = Logger::new(
+            constructor_options.log_queries,
+            log_level,
+            Box::new(log_callback),
+            false,
+        );
+
+        Ok(Self {
+            error_string: RwLock::new(CString::new("").unwrap()),
+            inner: RwLock::new(Inner::Builder(builder)),
+            base_path,
+            logger,
+        })
+    }
+
+    pub async fn connect(&self, trace: *const c_char) -> Result<(), Error> {
+        if let Some(base_path) = self.base_path.as_ref() {
+            env::set_current_dir(Path::new(&base_path)).expect("Could not change directory");
+        }
```

Comment:
The readme explains some of the quirks of this target

https://github.com/prisma/prisma-engines/pull/4774#diff-d54b9b218899f7be51fa6e28521b5bf70013c2f8cc776516f9ec5c6b45408a1eR24

In this case, it is because the file system on iOS (and the latests versions of Android) is sandboxed so you cannot access outside a specific folder. This value is also calculated on runtime, so there is no way to specify it on the schema without placeholders.

I tried the placeholder route, but could not get it to work. This was the easiest way to get it to work, without doing weird changes to the internals of the query engine.

Reply:
Got it, thanks for the explanation. Could you add a concise comment in the code about it, with a reference to the README for more info?